WhateverGreen/kern_igfx.cpp

```diff
@@ -651,6 +721,585 @@ int IGFX::wrapReadAUX(void *that, IORegistryEntry *framebuffer, uint32_t address
+	
+	// Retrieve the user preference
+	LSPCON *lspcon = nullptr;
+	auto pmode = framebufferLSPCONGetPreferredMode(index);
+	
+	// Guard: Check whether this framebuffer connector has an onboard LSPCON chip
+	if (!framebufferHasLSPCON(index)) {
+		// No LSPCON chip associated with this connector
+		DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] No LSPCON chip associated with this framebuffer.", index);
+		goto org;
+	}
+	
+	// Guard: Check whether the LSPCON driver has already been initialized for this framebuffer
+	if (framebufferHasLSPCONInitialized(index)) {
+		// Already initialized
+		lspcon = framebufferGetLSPCON(index);
+		DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] LSPCON driver (at 0x%llx) has already been initialized for this framebuffer.", index, lspcon);
+		// Confirm that the adapter is running in preferred mode
+		if (lspcon->setModeIfNecessary(pmode) != kIOReturnSuccess) {
+			SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] The adapter is not running in preferred mode. Failed to update the mode.", index);
+		}
+		// Wake up the native AUX channel if PCON mode is preferred
+		if (pmode == LSPCON::Mode::ProtocolConverter) {
+			if (lspcon->wakeUpNativeAUX() != kIOReturnSuccess)
+				SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] Failed to wake up the native AUX channel.", index);
+		}
+		goto org;
+	}
+	
+	// User has specified the existence of onboard LSPCON
+	// Guard: Initialize the driver for this framebuffer
+	lspcon = new LSPCON(that, framebuffer, displayPath);
+	if (lspcon == nullptr) {
+		SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] Failed to initialize the LSPCON driver. Insufficient memory.", index);
+		goto org;
+	}
+	
+	// Guard: Attempt to probe the onboard LSPCON chip
+	if (lspcon->probe() != kIOReturnSuccess) {
+		SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] Failed to probe the LSPCON adapter.", index);
+		delete lspcon;
+		lspcon = nullptr;
+		SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] Abort the LSPCON driver initialization.", index);
+		goto org;
+	}
+	DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] LSPCON driver has detected the onboard chip successfully.", index);
+	
+	// LSPCON driver has been initialized successfully
+	framebufferSetLSPCON(index, lspcon);
+	DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] LSPCON driver has been initialized successfully.", index);
+	
+	// Guard: Set the preferred adapter mode if necessary
+	if (lspcon->setModeIfNecessary(pmode) != kIOReturnSuccess) {
+		SYSLOG("igfx", "SC: GetDPCDInfo() Error: [FB%d] The adapter is not running in preferred mode. Failed to set the %s mode.", index,  LSPCON::getModeString(pmode));
+	}
+	DBGLOG("igfx", "SC: GetDPCDInfo() DInfo: [FB%d] The adapter is running in preferred mode [%s].", index, LSPCON::getModeString(pmode));
+	
+#ifdef DEBUG
+	framebuffer->setProperty("fw-framebuffer-has-lspcon", framebufferHasLSPCON(index));
+	framebuffer->setProperty("fw-framebuffer-preferred-lspcon-mode", LSPCON::getModeValue(pmode), 8);
+#endif
+	
+org:
```

Comment:
Are you sure this goto makes good sense here? I would say make some inline function in the header if necessary and make the above goto's invoke that function. While goto kind of does make sense for cleanup, here I see no justified reason for its use.

Reply:
Not at all. Previously, there were just a few lines of code in this function, but later I added functionalities like setting customizable mode, etc. Now I have decided to refactor this method. Please take a look at `framebufferSetupLSPCON()`.